lib/reply.js

```diff
@@ -140,27 +143,44 @@ function wrapOnSendEnd (err, reply, payload) {
 }
 
 function onSendEnd (reply, payload) {
+  var res = reply.res
+  var statusCode = res.statusCode
+
   if (payload === undefined || payload === null) {
     reply.sent = true
-    reply.res.end()
+
+    // according to https://tools.ietf.org/html/rfc7230#section-3.3.2
+    // we cannot send a content-length for 304 and 204, and all status code
+    // < 200.
+    if (statusCode >= 200 && statusCode !== 204 && statusCode !== 304) {
+      reply._headers['content-length'] = '0'
```

Comment:
> A server MAY send a Content-Length header field in a response to a HEAD request (Section 4.3.2 of [RFC7231]); a server MUST NOT send Content-Length in such a response unless its field-value equals the decimal number of octets that would have been sent in the payload body of a response if the same request had used the GET method.

If a user sends a 200 response to a HEAD request, this will most likely be the wrong Content-Length.

Reply:
Except this is within the condition that the payload is either `undefined` or `null`. Thus, there isn't a content length.

Reply:
what should it be in that case? Should we check if there is a content-length set before setting it to zero?

Reply:
I’m completely lost. This code passes all our tests in master without changes.

Reply:
@mcollina I believe you are correct with the code as it is. Look at the full context:

https://github.com/fastify/fastify/blob/578517d46d9edd90d11f157c5169113796019f0d/lib/reply.js#L149-L163

The payload size should always be 0 where you are explicitly setting the content length because a payload does not exist.

Reply:
@jsumners If it's a HEAD request, the user may decide to only set the headers and not send the payload itself (even though it would be stripped by Node). Users can do this to avoid spending time generating a payload.

@mcollina In that case the Content-Length should not be set if it was not already set by a user.

Reply:
It's okay. I'm not going to block this.

Reply:
It's still appropriate to send `Content-length: 0` in that case because it is true. Headers do not count.